### ISSUE TITLE
Steinhafels - Mattress pad rules

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2447]
_Imported from trac ticket 2447.
Created by **sjohnson**
Opened in trac: Aug 15, 2013, 4:10:32 PM
Last modified in trac: Oct 25, 2013, 9:48:23 AM_
***
If a slsp is writing a mattress order, they need to select one of the following on the order:
1. Mattress Pad (must purchase to received 120-night guarantee) 
2. Customer Waived 
3. As-Is Mattress 
4. Final Selection

See attached screen shot for the current pop-up in our POS system. All of the above options are SKUs that are entered on the order. And items 2-4 also add an auto comment (see below for examples).

Customer Waived Option = adds SLEEPWAVD sku and adds "SKU: 979123482 is SLEEPWAVD" in comments
As-Is Mattress Option = adds 999888880 sku and adds "SKU: 979546729 is AS-IS/NO SERVICE/HOLDS" in comments
Final Selection Option = adds FINLSELCT sku and adds "SKU: 939357718 is FINLSELCT" in comments

We add the comments section b/c it is a 1:1 ratio. If a customer is buying multiple mattresses on one order, they need a "reason or matt pad" for each mattress on the order.

What can we do in Loft to duplicate this scenario?
___
Please add the following as a rule to Loft. Here is the detail of what we look at (but you might be able to change or modify if you see something better):
* check if a mattress is on the order. A Mattress is defined as minor code = 930, 933, 935, 937, 939, 942, 944, 946, 950, 953, 955, 957, 959, 962, 966, 970, 973, 975, 979, 982, 987, 989 
* count how many mattresses are on the order 
* count how many FINLSELCT and SLEEPWAVD skus are on the order 
* count how many Pads are on the order. And to determine if it is a Pad, in the Finish Field the word "Cover" must be present. 
* the above needs to equal each other. If it does not equal, then show message requiring one of those be added.

***
* kbower attached [untitled-part.png](https://trac.retailarchitects.com/trac/attachment/ticket/2447/untitled-part.png) on Sep 30, 2013, 1:36:48 PM

**Comment by kentbower** [comment]
_**sjohnson** trac comment on Sep 20, 2013, 2:17:41 PM:_

Normally for this kind of rule, when more than 1 of the same SKU is present, we would force the same guarantee for all occurrences. But we won't do that here.  Need comment added for lines with sleep waved or finselect.  No comment needed for ones that have Pads.

**Comment by kentbower** [comment]
_**kbower** trac comment on Sep 30, 2013, 1:36:48 PM:_


[[Image(untitled-part.png)]]

**Comment by kentbower** [comment]
_**kbower** trac comment on Oct 1, 2013, 2:08:51 PM:_

Can you find out if SLEEPWAVD, FINLSELCT and 999888880 can only be linked to certain products (being an error to try to link them to invalid products) and, if so, what the rules are for each?

As 999777772 is a FAB product, it can only be added as already specified by the value product (treatable) framework.

**Comment by kentbower** [comment]
_**kbower** trac comment on Oct 2, 2013, 8:52:59 AM:_

Doug/Denise:

Our specifications given to us for mattress pad rules are as follows:

 * check if a mattress is on the order. A Mattress is defined as minor
   code = 930, 933, 935, 937, 939, 942, 944, 946, 950, 953, 955, 957,
   959, 962, 966, 970, 973, 975, 979, 982, 987, 989
 * count how many mattresses are on the order
 * count how many FINLSELCT and SLEEPWAVD skus are on the order
 * count how many Pads are on the order. And to determine if it is a
   Pad, in the Finish Field the word "Cover" must be present.
 * the above needs to equal each other. If it does not equal, then show
   message requiring one of those be added.


In addition to looking for a better definition of pad (previous email), there are some concerns here:
If the count of mattresses > count of options, there is no ambiguity: they need to make another choice, no problem there.
The problem comes when the count of pads + FINLSELCT + SLEEPWAVD + 999888880 > count of mattresses.  In this case, if we put up an error indicating they have chosen too many options given the count of mattresses, then no one will be able to sell a pad by itself without also buying a mattress (unless you create a "PAD ONLY" mattress sku).  Further, if FINLSELCT, SLEEPWAVD or 999888880 are used for other SKUs besides mattresses, then the design also falls apart.
On the other hand, if we allow pads + FINLSELCT + SLEEPWAVD + 999888880 > mattresses, then when a mattress is voided nothing will force the user to also void one of the choices (this might be fine, since the SKUs are captured in the comments).

Let us know how you would like us to proceed.  For now, I am assuming it is only an error if there are more mattresses than options, no error if other way around.

**_Keep it as this assumption_**

Thanks!
Kent

P.S. Also, to avoid hard coding the comment (so the comment code can work the same for all), I assumed it is best to use the vendorsku of the "special" skus.  The way your VSN data are currently, this means the comments will look like this:
SKU: 123456XYZ is NON-RETURNABLE ITEM
SKU: 123456XYZ is AS-IS/NO SERVICE/HOLDS
SKU: 123456XYZ is FINAL SELECTION
SKU: 123456XYZ is FACTORY WARRANTY ONLY

**_No objection_**

If this isn't sufficient, let me know.  Also, if these comments can't have CMNT_TYPE null (must be 'S'), let me know.

**_Want NULL_**

**Comment by kentbower** [comment]
_**kbower** trac comment on Oct 2, 2013, 9:28:45 AM:_

Waiting on Steinhafels for cleaner definition of mattress pads and other questions listed in this ticket.

# Still to implement
 * ~~related items for pads ~~
 * ~~related items for frames~~ ([#2529](2529))
 * ~~better definition of pads~~

**Comment by i-rabot** [comment]
_**kbower** trac status change on Oct 2, 2013, 9:28:45 AM:_ **waiting**

**Comment by i-rabot** [comment]
_**kbower** trac comment on Oct 2, 2013, 10:03:54 AM:_

(Changeset [[4602]](../commit/ba4b38d98fdc0b5e28164ef4d9a78487c9a58ff8))
* rudimentary Steinhafels mattress rules implementation, see #2447 #stn

**Comment by i-rabot** [comment]
_**sjohnson** trac comment on Oct 3, 2013, 9:44:10 AM:_

Doug,
Would grade be a reliable means of identifying Mattress size?
We could almost get it from the below data except 982904851 is designated as a "CALIFORNIA ING MATTRESS"

```
SQL> select count(*), grade from products where lower(grade) like '%mat%' and noticeid is null and lower(finish) not like '%cover%' and lower(grade) not like '%protector%' group by grade;

  COUNT(*) GRADE
---------- ------------------------------
         1 (2)TXL BASES/KING MATTRESS
         1 CALIFORNIA ING MATTRESS
         1 CALIFORNIA KING MATTRES
         60 CALIFORNIA KING MATTRESS
         71 FULL MATTRESS
         1 FULL MISMATCH MATTRESS
         1 FULL SIZE MATTRESS
         1 KING MATTRES
         79 KING MATTRESS
         1 QUEEN ADJ BASE AND MATTRESS
         1 QUEEN MATTRES
         79 QUEEN MATTRESS
         1 QUEEN MISMATCH MATTRESS
         1 SPLIT CAL KING MATTRESS
         1 TWIN  MATTRESWS
         1 TWIN EXTRA LONG MATTRESS
         54 TWIN MATTRESS
         1 TWIN MISMATCH MATTRESS
         2 TWIN S-LONG MATTRESS
         73 TWIN X-LONG MATTRESS
         1 TXIN X-LONG MATTRESS
         1 TXL ADJ BASE AND MATTRESS

22 rows selected.
```
___
Hi Scott,
Yes,  I checked and did no see any other than the one you pointed out that would be a problem.
Doug

**Comment by i-rabot** [comment]
_**kbower** trac comment on Oct 3, 2013, 10:13:51 AM:_

Replying to [sjohnson](comment:12):

It's just so ugly and relies on future purchasers correctly filling that in.  I think we make a product property for this as well.

**Comment by i-rabot** [comment]
_**sjohnson** trac comment on Oct 3, 2013, 2:13:11 PM:_

The data setup is complete on Steinhafels Train.  To get Related items when a Mattress is selected, match the custom property 'MattSize' on the Mattress to the Custom property 'Pad' on the mattress pads.

**Comment by i-rabot** [comment]
_**kbower** trac comment on Oct 3, 2013, 2:31:45 PM:_

(Changeset [[4614]](../commit/0049b2f63a946f81f54482b878c1dd717dda6d98))
* re #2447, add mattress pads to related items query (relies on Loft Custom Property data setup) #stn

**Comment by i-rabot** [comment]
_**sjohnson** trac comment on Oct 4, 2013, 4:05:07 PM:_

Scott,
Currently in our POS, we have FRAMES hard-coded in a pop-up box as they seldom change.  You are referring to the over-crowding by the related items in Loft, correct?  I will think if there is any easier way to show these items in Loft.  But I'd rather show box and pads first and if room then frames.  Let me toss it around this weekend and get back to you on Monday.
Denise

**Comment by i-rabot** [comment]
_**kbower** trac comment on Oct 9, 2013, 1:41:55 PM:_

(Changeset [[4642]](../commit/b04d2005c87271b3e645d23be5ad1ca2875a0bf3))
* refs #2447 & [#2483](2483), reorder Related Items #stn

**Comment by i-rabot** [comment]
_**kbower** trac comment on Oct 9, 2013, 4:52:31 PM:_

Only thing remaining to close this ticket is better definition of mattress pad than "COVER" in finish field.

Otherwise, closed.

**Comment by i-rabot** [comment]
_**kbower** trac status change on Oct 9, 2013, 4:52:31 PM:_ **waiting**

**Comment by i-rabot** [comment]
_**kbower** trac comment on Oct 11, 2013, 10:14:18 AM:_

Denise,
Right now we are using the definition you provided to identify Matt Pads ie "cover" in the finish.  This returns 38 items as shown below.  We are using this for the check to make sure there are at least as many of those as there are mattresses.  This looks like it could lead to undesirable results with things like WARREN DINING CHAIR in the list.  Do you think the first list below of 29 items is a better list or do you have any adjustments that should be made?
thanks,
Scott

```
SQL> select productid, vendor, vendorsku, description, finish from products where lower(finish) like '%cover%' and noticeid is null and vendorsku like '120NT%' order by finish;

PRODUCTID VEND VENDORSKU                      DESCRIPTION                    FINISH
--------- ---- ------------------------------ ------------------------------ ------------------------------
9999SQ889 PRTC 120NT GUARANTEE PAD            P-3080 SPLT QN COVER           30x80 MT COVER
964CKTN90 PRTC 120NT GUARANTEE                TEN0159 CK-PD                  CALIFORNIA KING MATTRESS COVER
964KCK711 PRTC 120NT GUARANTEE                KCLP0159 CK KIT                CALIFORNIA KING MATTRESS COVER
964AZCK94 PRTC 120NT GUARANTEE                BOM1813 CK ENC.                CALIFORNIA KING MATTRESS COVER
964032800 TMPR 120NT GUARANTEE                45703280 CK-PD                 CALIFORNIA KING MATTRESS COVER
999CK0152 PRTC 120NT GUARANTEE PAD            P0159 CK-PD                    CALIFORNIA KING MATTRESS COVER
999BAS465 PRTC 120NT GUARANTEE                BAS0128 FL-PD                  FULL MATTRESS COVER
964FTN221 PRTC 120NT GUARANTEE                TEN0128 FL-PD                  FULL MATTRESS COVER
999F01280 PRTC 120NT GUARANTEE PAD            P-0128 FL-PD                   FULL MATTRESS COVER
964AZFL03 PRTC 120NT GUARANTEE                BOM1309 FL ENC.                FULL MATTRESS COVER
999BAS667 PRTC 120NT GUARANTEE                BAS0142 KG-PD                  KING MATTRESS COVER
964AZKG10 PRTC 120NT GUARANTEE                BOM1713 KG ENC.                KING MATTRESS COVER
964KGK359 PRTC 120NT GUARANTEE                KCLP0142 KG KIT                KING MATTRESS COVER
999K01420 PRTC 120NT GUARANTEE PAD            P-0142 KG-PD                   KING MATTRESS COVER
964KTN601 PRTC 120NT GUARANTEE                TEN0142 KG-PD                  KING MATTRESS COVER
999BAS504 PRTC 120NT GUARANTEE                BAS0135 QN-PD                  QUEEN MATTRESS COVER
964QTN451 PRTC 120NT GUARANTEE                TEN0135 QN-PD                  QUEEN MATTRESS COVER
999Q01357 PRTC 120NT GUARANTEE PAD            P-0135 QN-PD                   QUEEN MATTRESS COVER
964AZQN45 PRTC 120NT GUARANTEE                BOM1513 QN ENC.                QUEEN MATTRESS COVER
964KQN723 PRTC 120NT GUARANTEE                KCLP0135 QN KIT                QUEEN MATTRESS COVER
999BAS338 PRTC 120NT GUARANTEE                BAS0111 TW-PD                  TWIN MATTRESS COVER
964TTN445 PRTC 120NT GUARANTEE                TEN0111 TW-PD                  TWIN MATTRESS COVER
964AZTW52 PRTC 120NT GUARANTEE                BOM1109 TW ENC.                TWIN MATTRESS COVER
999T01113 PRTC 120NT GUARANTEE PAD            P-0111 TW-PD                   TWIN MATTRESS COVER
999BASTX5 PRTC 120NT GUARANTEE                BAS0197 TXL-PD                 TWIN X-LONG MATTRESS COVER
964TXTN97 PRTC 120NT GUARANTEE                TEN0197 TXL-PD                 TWIN X-LONG MATTRESS COVER
964AZTX39 PRTC 120NT GUARANTEE                BOM1209 TXL ENC.               TWIN X-LONG MATTRESS COVER
999TX0197 PRTC 120NT GUARANTEE PAD            P-0197 TXL-PD                  TWIN X-LONG MATTRESS COVER
964KTX598 PRTC 120NT GUARANTEE                KCLP0197 TXL KIT               TWIN X-LONG MATTRESS COVER

29 rows selected.

SQL> select productid, vendor, vendorsku, description, finish from products where lower(finish) like '%cover%' and noticeid is null order by finish;

PRODUCTID VEND VENDORSKU                      DESCRIPTION                    FINISH
--------- ---- ------------------------------ ------------------------------ ------------------------------
104020158 HILS 4527-831 MILAN BLK/COPPER      30" SWIVEL BARSTOOL            30" BROWN FAUX SUEDE COVER
104382128 HILS 4324-831 MIDTOWN BLK 2TONE     30" SWIVEL VINYL BARSTOOL      30" STONE FAUX SUEDE COVER
9999SQ889 PRTC 120NT GUARANTEE PAD            P-3080 SPLT QN COVER           30x80 MT COVER
999CK0152 PRTC 120NT GUARANTEE PAD            P0159 CK-PD                    CALIFORNIA KING MATTRESS COVER
964CKTN90 PRTC 120NT GUARANTEE                TEN0159 CK-PD                  CALIFORNIA KING MATTRESS COVER
964KCK711 PRTC 120NT GUARANTEE                KCLP0159 CK KIT                CALIFORNIA KING MATTRESS COVER
964032800 TMPR 120NT GUARANTEE                45703280 CK-PD                 CALIFORNIA KING MATTRESS COVER
964AZCK94 PRTC 120NT GUARANTEE                BOM1813 CK ENC.                CALIFORNIA KING MATTRESS COVER
650415702 DOVE DOV157                         WARREN DINING CHAIR            COTTON SLIPCOVER - NAVY
363348383 LFST U1111S-60L-XXXX-SQM-PIBL       PALERMO LAF SOFA 80X41.5X30"H  COVER: PISA/BLACK
374348386 LFST U1111S-10S-XXXX-SQM-PIBL       PALERMO CKTL OTTOMAN 63X31X19  COVER: PISA/BLACK
364348384 LFST U1111S-70R-XXXX-SQM-PIBL       PALERMO RAF CHAISE 41.5X94X30  COVER: PISA/BLACK
363348371 LFST U1111S-60L-XXXX-SQM-PIPW       PALERMO LAF SOFA 80X41.5X30"H  COVER: PISA/PLAIN WHITE
364348372 LFST U1111S-70R-XXXX-SQM-PIPW       PALERMO RAF CHAISE 41.5X94X30  COVER: PISA/PLAIN WHITE
374348374 LFST U1111S-10S-XXXX-SQM-PIPW       PALERMO CKTL OTTOMAN 63X31X19  COVER: PISA/PLAIN WHITE
999F01280 PRTC 120NT GUARANTEE PAD            P-0128 FL-PD                   FULL MATTRESS COVER
999BAS465 PRTC 120NT GUARANTEE                BAS0128 FL-PD                  FULL MATTRESS COVER
964FTN221 PRTC 120NT GUARANTEE                TEN0128 FL-PD                  FULL MATTRESS COVER
964AZFL03 PRTC 120NT GUARANTEE                BOM1309 FL ENC.                FULL MATTRESS COVER
999K01420 PRTC 120NT GUARANTEE PAD            P-0142 KG-PD                   KING MATTRESS COVER
999BAS667 PRTC 120NT GUARANTEE                BAS0142 KG-PD                  KING MATTRESS COVER
964AZKG10 PRTC 120NT GUARANTEE                BOM1713 KG ENC.                KING MATTRESS COVER
964KGK359 PRTC 120NT GUARANTEE                KCLP0142 KG KIT                KING MATTRESS COVER
964KTN601 PRTC 120NT GUARANTEE                TEN0142 KG-PD                  KING MATTRESS COVER
999Q01357 PRTC 120NT GUARANTEE PAD            P-0135 QN-PD                   QUEEN MATTRESS COVER
964KQN723 PRTC 120NT GUARANTEE                KCLP0135 QN KIT                QUEEN MATTRESS COVER
964QTN451 PRTC 120NT GUARANTEE                TEN0135 QN-PD                  QUEEN MATTRESS COVER
999BAS504 PRTC 120NT GUARANTEE                BAS0135 QN-PD                  QUEEN MATTRESS COVER
964AZQN45 PRTC 120NT GUARANTEE                BOM1513 QN ENC.                QUEEN MATTRESS COVER
999T01113 PRTC 120NT GUARANTEE PAD            P-0111 TW-PD                   TWIN MATTRESS COVER
999BAS338 PRTC 120NT GUARANTEE                BAS0111 TW-PD                  TWIN MATTRESS COVER
964AZTW52 PRTC 120NT GUARANTEE                BOM1109 TW ENC.                TWIN MATTRESS COVER
964TTN445 PRTC 120NT GUARANTEE                TEN0111 TW-PD                  TWIN MATTRESS COVER
999TX0197 PRTC 120NT GUARANTEE PAD            P-0197 TXL-PD                  TWIN X-LONG MATTRESS COVER
999BASTX5 PRTC 120NT GUARANTEE                BAS0197 TXL-PD                 TWIN X-LONG MATTRESS COVER
964TXTN97 PRTC 120NT GUARANTEE                TEN0197 TXL-PD                 TWIN X-LONG MATTRESS COVER
964KTX598 PRTC 120NT GUARANTEE                KCLP0197 TXL KIT               TWIN X-LONG MATTRESS COVER
964AZTX39 PRTC 120NT GUARANTEE                BOM1209 TXL ENC.               TWIN X-LONG MATTRESS COVER

38 rows selected.
```
-- 

Scott Johnson

**Comment by i-rabot** [comment]
_**kbower** trac comment on Oct 25, 2013, 9:45:16 AM:_

Scott,
I received final confirmation from our Mattress Dept that 120NT in the VSN field will be an excellent way to determine if its a valid mattress pad when purchasing a mattress.

If you have further questions, let me know.

Denise

**Comment by i-rabot** [comment]
_**kbower** trac comment on Oct 25, 2013, 9:48:23 AM:_

(Changeset [[4716]](../commit/66b1689a3bfb5217f5df777aab7cfd19df42d827))
* closes #2447, Steinhafels - Mattress pad rules. Defines mattress pads as COVER in finish and 120NT at start of vendorsku #stn

**Comment by i-rabot** [comment]
_**kbower** trac status change on Oct 25, 2013, 9:48:23 AM:_ **closed**
